lucene/expressions/src/java/org/apache/lucene/expressions/SimpleBindings.java

```diff
@@ -96,24 +90,51 @@ public DoubleValuesSource getDoubleValuesSource(String name) {
       case SCORE:
         return DoubleValuesSource.SCORES;
       default:
-        throw new UnsupportedOperationException(); 
+        throw new UnsupportedOperationException();
     }
   }
   
-  /** 
-   * Traverses the graph of bindings, checking there are no cycles or missing references 
-   * @throws IllegalArgumentException if the bindings is inconsistent 
+  @Override
+  public DoubleValuesSource getDoubleValuesSource(String name) {
+    if (map.containsKey(name) == false) {
+      throw new IllegalArgumentException("Invalid reference '" + name + "'");
+    }
+    return map.get(name).apply(this);
+  }
+
+  /**
+   * Traverses the graph of bindings, checking there are no cycles or missing references
+   * @throws IllegalArgumentException if the bindings is inconsistent
    */
   public void validate() {
```

Comment:
Have you considered returning the map, or an immutable view on it, so that callers can use this to enumerate all the dependencies? In a similar framework, I've found this to be pretty helpful for analyzing query patterns. It's also nice to know if the same name occurs multiple times in the dependency tree; maybe one should cache its value in that case.

Reply:
Cacheing is a whole other conversation, which I think is related to the stuff that @mkhludnev is working on around grouping (in that we could plausibly have multiple references to the same iterator all moving in lockstep, where at the moment we pull separate iterators for each reference).  But I think that's for a follow-up really, this issue is just a bit of refactoring to improve type safety.